runtime/src/iree/hal/drivers/cuda/tracing.c

```diff
@@ -169,98 +195,188 @@ void iree_hal_cuda_tracing_context_free(
+    events = next;
+    context->event_submitted_list_head = events;
+  }
+  IREE_TRACE_ZONE_APPEND_VALUE_I64(z0, (int64_t)read_query_count);
+  IREE_TRACE_ZONE_END(z0);
+  iree_slim_mutex_unlock(&context->event_mutex);
+}
+
+void iree_hal_cuda_tracing_notify_submitted(
+    iree_hal_cuda_tracing_context_t* context,
+    iree_hal_cuda_tracing_context_event_t** event_list_begin,
+    iree_hal_cuda_tracing_context_event_t** event_list_end) {
+  if (!context) return;
+  iree_slim_mutex_lock(&context->event_mutex);
+
+  IREE_ASSERT_ARGUMENT(event_list_begin);
+  IREE_ASSERT_ARGUMENT(event_list_end);
 
-    context->query_tail += read_query_count;
-    if (context->query_tail >= context->query_capacity) {
-      context->query_tail = 0;
+  if (!*event_list_begin) {
+    iree_slim_mutex_unlock(&context->event_mutex);
+    return;
+  }
+
+  iree_hal_cuda_tracing_context_event_t* evt = *event_list_begin;
+  while (evt) {
+    evt = evt->next_in_cb;
+  }
+
+  if (!context->event_submitted_list_head) {
+    context->event_submitted_list_head = *event_list_begin;
+    context->event_submitted_list_tail = *event_list_begin;
+    iree_slim_mutex_unlock(&context->event_mutex);
+    return;
+  }
+
+  context->event_submitted_list_tail->next_submission = *event_list_begin;
+  context->event_submitted_list_tail = *event_list_begin;
+  iree_slim_mutex_unlock(&context->event_mutex);
+}
+
+void iree_hal_cuda_tracing_free(
+    iree_hal_cuda_tracing_context_t* context,
+    iree_hal_cuda_tracing_context_event_t** event_list_begin,
+    iree_hal_cuda_tracing_context_event_t** event_list_end) {
+  if (!context) return;
+  iree_slim_mutex_lock(&context->event_mutex);
+
+  IREE_ASSERT_ARGUMENT(event_list_begin);
+  IREE_ASSERT_ARGUMENT(event_list_end);
+
+  if (!*event_list_begin) {
+    iree_slim_mutex_unlock(&context->event_mutex);
+    return;
+  }
+
+  // If this event list has never been submitted,
+  // we still need to add values to the timeline,
+  // otherwise tracy will not behave correctly.
+  if ((!(*event_list_begin)->next_submission)) {
+    iree_hal_cuda_tracing_context_event_t* event = *event_list_begin;
+    while (event) {
+      uint32_t query_id = (uint32_t)(event - &context->event_pool[0]);
+      iree_tracing_gpu_zone_notify(context->id, query_id, 0);
+      event = event->next_in_cb;
     }
   }
 
-  IREE_TRACE_ZONE_END(z0);
+  if (!context->event_freelist_head) {
+    context->event_freelist_head = *event_list_begin;
+    iree_slim_mutex_unlock(&context->event_mutex);
+    return;
+  }
+  (*event_list_begin)->next_submission = NULL;
+  (*event_list_end)->next_in_cb = context->event_freelist_head;
+  context->event_freelist_head = *event_list_begin;
+
+  *event_list_begin = NULL;
+  *event_list_end = NULL;
+  iree_slim_mutex_unlock(&context->event_mutex);
 }
 
 static uint16_t iree_hal_cuda_stream_tracing_context_insert_query(
-    iree_hal_cuda_tracing_context_t* context, CUstream stream) {
+    iree_hal_cuda_tracing_context_t* context,
```

Comment:
Add some documentation to this function given now it's not as straightforward as before? Mostly need to be clear what we are updating to in the function parameters. Something like "Grabs the next available event from |context|'s freelist, records it into |stream|, and add to the |event_list|.